File: 02/02.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define N_GAMES 2500
// <c>_<c>\n\0
#define LINE_LEN 5

char* read_data(const char* file_name, size_t n_games, size_t line_len) {
    FILE* fp = fopen(file_name, "r");
    if (fp == NULL) {
        perror(file_name);
        return NULL;
    }

    char* games = malloc(sizeof(char) * (n_games * 2));
    if (games == NULL) {
        fclose(fp);
        return NULL;
    }

    char* buf = malloc(sizeof(char) * line_len);
    if (buf == NULL) {
        fclose(fp);
        free(games);
        return NULL;
    }

    int i = 0;
    while (fgets(buf, line_len, fp) != NULL) {
        games[i++] = buf[0];
        games[i++] = buf[2];
    }
    fclose(fp);
    free(buf);

    return games;
}

int part1(char* data, size_t n_games) {
    char diff = 'X' - 'A';
    char offset = 'W';

    int score = 0;
    for (int i = 0; i < n_games * 2; i++) {
        char opp = data[i++];
        char you = data[i];

        if (opp + diff == you) {
            score += 3;
        } else if ((opp == 'A' && you == 'Y') || (opp == 'B' && you == 'Z') ||
                   (opp == 'C' && you == 'X')) {
            score += 6;
        }

        score += you - offset;
    }

    return score;
}

int part2(char* data, size_t n_games) {
    int score = 0;
    for (int i = 0; i < n_games * 2; i++) {
        char opp = data[i++];
        char you = data[i];

        if (you == 'X') {  // loss
            if (opp == 'A')
                score += 3;
            else if (opp == 'B')
                score += 1;
            else
                score += 2;
        } else if (you == 'Y') {  // draw
            score += 3 + (opp - 'A' + 1);
        } else {  // win
            score += 6;
            if (opp == 'A')
                score += 2;
            else if (opp == 'B')
                score += 3;
            else
                score += 1;
        }
    }

    return score;
}

unsigned long timestamp_nano() {
    struct timespec time;
    timespec_get(&time, TIME_UTC);
    return time.tv_nsec;
}

int main() {
    unsigned long start = timestamp_nano();

    char* data = read_data("02/input.txt", N_GAMES, LINE_LEN);
    if (data == NULL) exit(EXIT_FAILURE);

    int p1 = part1(data, N_GAMES);
    printf("part1: %d\n", p1);

    int p2 = part2(data, N_GAMES);
    printf("part2: %d\n", p2);

    free(data);

    unsigned long end = timestamp_nano();
    printf("time: %.1f??s\n", (end - start) / 1000.0);

    exit(EXIT_SUCCESS);
}

File: 02/02.rs
use std::fs;
use std::time::Instant;

fn main() {
    let start = Instant::now();

    let games: Vec<(char, char)> = fs::read_to_string("02/input.txt")
        .expect("File not found")
        .lines()
        .map(|line| {
            let mut chars = line.chars();
            let c1: char = chars.next().unwrap();
            let c2: char = chars.nth(1).unwrap();

            (c1, c2)
        })
        .collect();

    println!("part1: {}", part1(&games));
    println!("part2: {}", part2(&games));

    println!("time: {:?}", start.elapsed());
}

fn part1(games: &[(char, char)]) -> i32 {
    let diff = b'X' - b'A';
    let offset = b'W';

    let mut score: i32 = 0;
    for &(opp, you) in games {
        if opp as u8 + diff as u8 == you as u8 {
            score += 3;
        } else if opp == 'A' && you == 'Y' || opp == 'B' && you == 'Z' || opp == 'C' && you == 'X' {
            score += 6;
        }
        score += (you as u8 - offset as u8) as i32;
    }
    score
}

fn part2(games: &[(char, char)]) -> i32 {
    let mut score: i32 = 0;
    for &(opp, you) in games {
        if you == 'X' {
            // loss
            if opp == 'A' {
                score += 3;
            } else if opp == 'B' {
                score += 1;
            } else {
                score += 2;
            }
        } else if you == 'Y' {
            // draw
            score += (3 + (opp as u8 - b'A' + 1)) as i32;
        } else {
            // win
            score += 6;
            if opp == 'A' {
                score += 2;
            } else if opp == 'B' {
                score += 3;
            } else {
                score += 1;
            }
        }
    }
    score
}
